pkg/csvcopy/csvcopy_test.go

```diff
@@ -255,3 +345,171 @@ func TestWriteReportProgress(t *testing.T) {
 	require.NoError(t, err)
 	assert.Equal(t, []interface{}{int32(24), "qased", 2.4}, results)
 }
+
+func TestFailedBatchHandler(t *testing.T) {
+	ctx := context.Background()
+
+	pgContainer, err := postgres.RunContainer(ctx,
+		testcontainers.WithImage("postgres:15.3-alpine"),
+		postgres.WithDatabase("test-db"),
+		postgres.WithUsername("postgres"),
+		postgres.WithPassword("postgres"),
+		testcontainers.WithWaitStrategy(
+			wait.ForLog("database system is ready to accept connections").
+				WithOccurrence(2).WithStartupTimeout(5*time.Second)),
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	t.Cleanup(func() {
+		if err := pgContainer.Terminate(ctx); err != nil {
+			t.Fatalf("failed to terminate pgContainer: %s", err)
+		}
+	})
+
+	connStr, err := pgContainer.ConnectionString(ctx, "sslmode=disable")
+	require.NoError(t, err)
+
+	conn, err := pgx.Connect(ctx, connStr)
+	require.NoError(t, err)
+	defer conn.Close(ctx)
+	_, err = conn.Exec(ctx, "create table public.metrics (device_id int, label text, value float8)")
+	require.NoError(t, err)
+
+	// Create a temporary CSV file
+	tmpfile, err := os.CreateTemp("", "example")
+	require.NoError(t, err)
+	defer os.Remove(tmpfile.Name())
+
+	// Write data to the CSV file
+	writer := csv.NewWriter(tmpfile)
+
+	data := [][]string{
+		// Batch 1
+		{"42", "xasev", "4.2"},
+		{"24", "qased", "2.4"},
+		// Batch 2
+		{"24", "qased", "2.4"},
+		{"24", "qased", "hello"},
+		// Batch 3
+		{"24", "qased", "2.4"},
+		{"24", "qased", "2.4"},
+	}
+
+	for _, record := range data {
+		if err := writer.Write(record); err != nil {
+			t.Fatalf("Error writing record to CSV: %v", err)
+		}
+	}
+
+	writer.Flush()
+	fs := &MockErrorHandler{}
+
+	copier, err := NewCopier(connStr, "metrics", WithColumns("device_id,label,value"), WithBatchSize(2), WithBatchErrorHandler(fs.HandleError))
+	require.NoError(t, err)
+	reader, err := os.Open(tmpfile.Name())
+	require.NoError(t, err)
+	result, err := copier.Copy(context.Background(), reader)
+	assert.NoError(t, err)
+	assert.EqualValues(t, 4, result.RowsRead)
+
+	require.Contains(t, fs.Files, 1)
+	assert.Equal(t, fs.Files[1].String(), "24,qased,2.4\n24,qased,hello\n")
+	require.Contains(t, fs.Errors, 1)
+	assert.EqualValues(t, fs.Errors[1].(*ErrAtRow).RowAtLocation(), 3)
+	assert.EqualValues(t, fs.Errors[1].(*ErrAtRow).BatchLocation.ByteOffset, 26)
+}
+
+type MockErrorHandler struct {
+	Files  map[int]*bytes.Buffer
+	Errors map[int]error
+}
+
+func (fs *MockErrorHandler) HandleError(batch batch.Batch, reason error) error {
```

Comment:
Why not just test with `BatchHandlerLog`. By doing this you're testing internals instead of behavior.
You could inspect the logging instead with something like https://stackoverflow.com/questions/44119951/how-to-check-a-log-output-in-go-test

Reply:
This is not testing internals. The ErrorHandler is part of the public interface when you use this as a package. 

What I should provably do is make sure I test all the fields, as right now I'm just testing that the buffer and the reason are correct.

Reply:
I think it'd be more useful to test with one of the builtin handlers BatchHandlerLog or BatchHandlerSaveToFile which also don't have their own coverage as far as I see